Add first unittests

johnny9 opened the pull request:
This adds an initial unittest against the bitcoinamount module. This is the simplest c++ module to test so its an easy first target. I think the Qt test framework is the best fit for the project and eventually I imagine GMock will be needed to mock out portions of bitcoin interfaces as well as more complex modules get tested. These changes update the CMake definitions and add a BUILD_APP_TESTS option to toggle building of the unittests. 

A new ci workflow is added as well to run the unittests (and eventually other testing that we have).

johnny9 committed:
Add a unittest against bitcoinamount

johnny9 committed:
Update README with instructions for test

johnny9 committed:
Add workflow for ci testing

johnny9 committed:
Update workflow depends to match README

johnny9 committed:
Remove cmake from macos depends in artifacts workflow

The macos runner already includes an up to date cmake. Trying to install it again can cause issues.

jarolrod commented:
ACK 16e86dddbdc069d126a5f2131f181cee05c80171

@hebasto can you give your opinion on the CMAKE toggle here

johnny9 committed:
test: add minimal unit and qml test targets

Add a shared C++ unit runner for bitcoinamount and a Qt QuickTest runner with one ValueInput QML test.

johnny9 committed:
ci, docs: install required Qt6 QML runtime modules for tests

johnny9 commented:
Squashed the ci fixes

johnny9 commented:
I introduced a small qml test as well. We will use Qtest to test the c++ modules as well as the qml

D33r-Gee commented:
tACK [236141e](https://github.com/bitcoin-core/gui-qml/commit/236141ea2db490114cd25976cc442be733467f06) on WSL Unbuntu 24.04.3

The binaries build without issues and the test returned "Passed"

Also the additions of the qml modules dependencies is helpful

```
ctest --test-dir build --output-on-failure

 Start 1: bitcoinqml_unit_tests
1/2 Test #1: bitcoinqml_unit_tests ............   Passed    0.04 sec
    Start 2: bitcoinqml_qmltests
2/2 Test #2: bitcoinqml_qmltests ..............   Passed    0.04 sec
```

hebasto commented on CMakeLists.txt:
```diff
@@ -129,4 +126,11 @@ target_link_libraries(bitcoin-core-app
+
+option(BUILD_APP_TESTS "Build unit tests for the app" ON)
+if(BUILD_APP_TESTS)
+  include(CTest)
+  enable_testing()
```
From `enable_testing` [docs](https://cmake.org/cmake/help/latest/command/enable_testing.html):
> This command is also automatically invoked when the [CTest](https://cmake.org/cmake/help/latest/module/CTest.html#module:CTest) module is included, except if the [BUILD_TESTING](https://cmake.org/cmake/help/latest/variable/BUILD_TESTING.html#variable:BUILD_TESTING) option is turned off.

```suggestion
  include(CTest)
```
